### PR TITLE
[New Blog Post] Orchestrating Events with Knative and Kogito

### LINKED ISSUE #3076: New blog post: Kogito and Knative Eventing integration example
Hi!

I'm currently working on [Kogito](https://kogito.kie.org/) project, which is a platform to build business applications. One of our features is the implementation of the [Serverless Workflow specification](https://serverlessworkflow.io/) (a CNCF sandbox project). In order to support eventing in the platform, we use Knative Eventing.

I'm writing a blog post about this integration introducing a [small example](https://github.com/kiegroup/kogito-examples/tree/master/serverless-workflow-functions-events-quarkus). Is it possible to post on Knative dev blog? I think it will be a nice addition to it since it shows how Knative can also be used in workflow orchestrations.

To make it easy for you, I can share a GDoc with the blog before sending the PR.

### PR DESCRIPTION
**Pull request by ricardozanini**
Fixes #3076

## Proposed Changes 

- A new entry for the Knative blog introducing the Serverless Workflow specification, Kogito and Knative Eventing

### CODE REVIEW

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:20]:
```suggestion
Kogito implements the [CNCF Serverless Workflow Project](https://github.com/serverlessworkflow/specification), which is a specification for defining workflow models that orchestrate event-driven, serverless applications. It focuses on defining a vendor-neutral, platform-independent, and declarative workflow model for orchestrating services that can be used across multiple cloud and container platforms. To date, the Serverless Workflow specification is a CNCF sandbox project.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:34]:
```suggestion
Here is an example of the patient onboarding workflow YAML, based on this specification:
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:26]:
```suggestion
To demonstrate how the Kogito workflow implementation works on event-driven architectures, we will use the [patient onboarding example](https://github.com/kiegroup/kogito-examples/tree/master/serverless-workflow-functions-events-quarkus). In this example, we simulate a workflow used in hospitals to onboard new patients and assign them to the correct doctor.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:71]:
```suggestion
This specification is based on three parts. The first part is the `states` array, which includes the workflow control flow logic. In this case, the state is an `event`, which receives the new patient cloud event, and performs the three functions mentioned previously.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:73]:
```suggestion
The second part of the spec is the `events` array, which contains event definitions that are based on the CloudEvent specification. There is a 1:1 mapping between how the events are defined in the workflows, and how they are represented with the CloudEvent format.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:32]:
```suggestion
The workflow starts after receiving a [CloudEvent](https://github.com/cloudevents/spec) object that contains patient information. Three functions are then called by the spec in a sequence which: (1) stores the patient information, (2) assigns the patient to a doctor based on their symptoms, and (3) schedules an appointment with the assigned doctor for that patient.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:75]:
```suggestion
The third part, the `functions` array, contains the function definitions that give more information to the runtime about how to execute the required services.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:81]:
```suggestion
Note that the specification allows for both JSON and YAML workflow formats. This example uses YAML, however JSON is considered equivalent and is also parsable by the runtime.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:87]:
```suggestion
In this example, we have also added the [Knative Kogito Eventing plugin](https://docs.jboss.org/kogito/release/latest/html_single/#con-knative-eventing_kogito-developing-process-services) to the project, which means that it can accept CloudEvent objects through HTTP on the root path. For example:
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:100]:
```suggestion
When sending this request to the service, a new instance of the workflow starts, and performs all of the operations defined in the spec.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:106]:
```suggestion
Based on this generated service, you can build [an image](https://docs.jboss.org/kogito/release/latest/html_single/#proc-kogito-deploying-on-kubernetes_kogito-deploying-on-openshift) to be deployed with the [Kogito Operator](https://docs.jboss.org/kogito/release/latest/html_single/#con-kogito-operator-and-cli_kogito-deploying-on-openshift) on a Kubernetes cluster with [Knative Eventing](https://knative.dev/docs/eventing/) installed. The Operator will create all the necessary Knative resources to configure this service and subscribe it to the Knative [broker](https://knative.dev/docs/eventing/broker/):
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:110]:
```suggestion
The Kogito Operator creates a Knative [Trigger](https://knative.dev/docs/eventing/triggers/) resource, that links the service and the broker together. In this example, it will filter events of type `new.patients.events`. This means that every time a new event of this type comes to the broker, it will be redirected to the Kogito service.
```

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:112]:
```suggestion
The same concept also applies to [events produced by the workflow engine](https://github.com/serverlessworkflow/specification/blob/0.5.x/specification.md#event-definition). In this case, the operator will create a Knative [SinkBinding](https://knative.dev/docs/eventing/samples/sinkbinding/) resource, and will bind it to the Knative broker. Each time an event is produced by the service, a CloudEvent representing it will be sent to the broker.
```

**Review by abrennan89 — CHANGES_REQUESTED** [review]
Some suggestions, but overall looks great!

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (ricardozanini) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:71]: Hi @abrennan89 many thanks for your review! :smile: 

This paragraph is about the workflow definition listed above. I think "This specification" won't reflect that, so the reader wouldn't associate the workflow and the description that follows it.

**Comment by ricardozanini** [comment]
@googlebot I fixed it.

**Comment by ricardozanini** [comment]
I had to rebase to fix the `cla` after accepting @abrennan89 suggestions. :(

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (tsurdilo) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:71]: Would it be more clear to start that sentence with
"The above workflow definition contains three mayor parts ...."
The entire workflow definition complies with the specification. Just trying to figure out what you think would be best wording for it.

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (tsurdilo) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:73]: Same here, I think this paragraph is specific to the actual workflow definition shown, and not the specification itself. 
From the spec perspective valid workflows that don't consume events would not define the "events" array", so this is specifically describing the example workflow in the doc.

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:5]: Note: Remember to add a date when we merge :slightly_smiling_face:

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:16]: Note: Add me when merging :slightly_smiling_face: anyone else who needs to sign off?

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:30]:
```suggestion
![Patient onboarding workflow representation](https://raw.githubusercontent.com/serverlessworkflow/specification/master/media/examples/example-patientonboarding.png)*Patient Onboarding workflow representation*
```

You can just make the changes yourself instead of accepting the suggestion if it causes you to need to rebase etc, but I'll add them as suggestions anyway so the change is highlighted.

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (abrennan89) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:71]: That's fine - could you please make the changes to `states` etc though, since they're literal fields so it seems like they should be formatted like `this` instead of "this"? I don't mind what way you choose to word this part though, and thanks for clarifying :)

**Review by abrennan89 — APPROVED** [review]
/hold

Just a few minor comments, thanks for adding the other changes :)

Once you're done and we have a date and reviewers added to the file I can go ahead and lgtm this to merge. Is there anyone else who needs to approve it first, SME wise?

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (ricardozanini) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:5]: Which date should I add? Maybe today?

**Comment by ricardozanini** [comment]
> Once you're done and we have a date and reviewers added to the file I can go ahead and lgtm this to merge. Is there anyone else who needs to approve it first, SME wise?

Thanks, @abrennan89! Well, I'm not aware of the approval process. I'd like to see an SME from the Knative Eventing community to take a look into this as well, if possible. From Kogito/SW specification side we already reviewed.

I just pushed the new version of the blog based on your comments, many thanks again! :smiley:

**Comment by abrennan89** [comment]
Assigned a few folks as reviewers to see if someone can take a look, let me post this in Slack too and see if we can get an approval.

**Comment by ricardozanini** [comment]
@macruzbar PR updated with your review. Mind taking a look at it? Especially in the conclusion section where I added the call to action you asked for. Many thanks for the feedback, I really appreciated.

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (lberk) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:105]: trailing ':'

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (lberk) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:107]: This line didn't seem to render properly for me in the preview

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (lberk) [inline blog/articles/event-drive-app-knative-eventing-kogito.md:111]: could we be explicit in what the `sinkbinding` subject is here?  If we're using yaml for this example, would it be possible to link to the yaml resource?

Comment (ricardozanini) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:111]: Hi @lberk ! Thanks for your review. The `sinkbinding` here is created automatically by the Kogito Operator once a workflow is deployed, so we don't have the yaml source (I could create an example thou). See: 
https://docs.jboss.org/kogito/release/latest/html_single/#con-knative-eventing_kogito-developing-process-services

I think we can link this doc here?

In this particular example, we only have the `trigger` to be created since we listening to the new patient event.

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (ricardozanini) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:107]: hmmm we need a new line here.

File: blog/articles/event-drive-app-knative-eventing-kogito.md
Comment (lberk) [reply blog/articles/event-drive-app-knative-eventing-kogito.md:111]: Hi! Fair enough, I think I was just trying to say 'I'm not sure what exactly the subject is' from the link you provided I think this diagram summarizes it (and if we could figure out how to communicate this somehow in the blog post, it'd be great!) https://docs.jboss.org/kogito/release/latest/html_single/kogito/bpmn/kogito-knative-impl-producing-event.png

Comment (ricardozanini) [reply blog/articles/event-drive-app-knative-eventing-kogito.md]: @lberk I pushed an update to the article incorporating this image and introducing it to this context, so readers can have a better view of the concept. I don't want to extend too much since this is an introductory article about this integration. I intend to keep writing more about these concepts in the future using this article as a reference. wdyt?

**Comment by ricardozanini** [comment]
@abrennan89 I've added @macruzbar and @lberk to the reviewers list. You guys tell me if we are good to go. :)

**Comment by lberk** [comment]
/lgtm

Thanks for putting this together @ricardozanini!
@abrennan89 from an Eventing side I think we're good, I'll leave the final `/unhold` to you though!

**Pull request merged**
